## Introduction to the data frame - Music Therapy

The post-pandemic era presents unprecedented challenges for individuals managing their mental health. Studies indicate that music has played a significant role in improving both physical and mental well-being (https://www.ncbi.nlm.nih.gov/pmc/articles/PMC8566759/). I am eager to explore the depth of music's influence on mental health and extract valuable insights from this study.

We want to identify what, if any, correlations exist between an individual's music taste and their self-reported mental health. Ideally, these findings could contribute to a more informed application of MT or simply provide interesting sights about the mind.

In [ ]:
#pip install numpy
#pip install pandas
#pip install matplotlib

In [1]:
# Ignore warnings
import warnings
warnings.filterwarnings("ignore")  # This is to ignore any warnings that might pop up during execution

In [2]:
# Basic libraries to manipulate data
import matplotlib.pyplot as plt  # Matplotlib for data visualization
import numpy as np  # Numpy for numerical computations
import pandas as pd  # Pandas for data manipulation

In [3]:
np.random.seed(42)  # To ensure all the probabilistic things are reproducible

## Background on Data

#### i. Dataset
For this analysis, we take the data from the Kaggle website (https://www.kaggle.com/datasets/catherinerasgaitis/mxmh-survey-results).

As per the dataset collector and owner, the survey encompassed the following parameters:
**Block 1: Music genres**
Respondents selected Never, Rarely, Sometimes, or Very frequently to indicate the frequency of their listening to 16 music genres.

**Block 2: Mental health**
Respondents rated Anxiety, Depression, Insomnia, and OCD (Obsessive-compulsive disorder) on a scale of 0 to 10, with 10 representing the highest frequency and severity.

#### ii. Scope & Procedures
This analysis is based on data collected from 736 respondents across the US. The survey, conducted from Aug. 27 to Nov. 9, 2022, included participants of any age or location. The questionnaire comprised a mix of open-ended and drop-down questions, designed to elicit relatively concise responses and encourage participation.

# On Relative Paths and Well-Structured Projects
As you'll often collaborate within a team, establishing order from the outset is a beneficial practice. This ensures code portability and facilitates bug fixing.

For reading data from files, it's advisable to use a relative path—i.e., a path originating from the project directory. This way, being in the correct directory ensures smooth execution.

Throughout this course, each project will have its own directory. At the top level, notebooks will reside, while all datasets will be stored in a directory named "data."

Consequently, we can define the relative path for data and effortlessly read files by appending the correct file name from there.

If you're using `google colab`, things get a bit trickier. You'll need to upload the data file  each time you want to access it in a separate session.

Alternatively, you can mount your Google Drive, head to your project directory, and then employ relative paths just like before.

In [1]:
# Mount your drive
# ! This will only work in Google Colab.
#from google.colab import drive
#drive.mount('/content/drive')

# Navigate to your project's directory
# ! This will vary based on each person's drive structure.
#import os
#os.chdir('/content/drive/My Drive/.../data/')

# Define the relative path within
#data_path = "./data/"

# Basic characteristics of the datasets

Now that we've completed the previous steps, let's take a snapshot of the datasets. Understanding the basic characteristics is crucial; without knowing what we have, we can't address the questions we aim to answer.

To kick things off, we'll import our tables into a Pandas dataframe.

In [4]:
# Specify the path to the datasets
data_path = "./data/"

# Specify the filenames of the datasets
survey_filename = "survey_music.csv"

# Read the CSV files and create backup copies
survey_df_data = pd.read_csv(data_path + survey_filename)

# Create working copies of the dataframes for analysis
survey_df = survey_df_data.copy()


Since this table has many rows, viewing it in its entirety won't give us the full picture. 
Various methods are available for a glimpse of the contents.

To view the start and end of the table, we can use `head` and `tail`. Alternatively, returning a sample gives a broader view, considering potential correlations in rows. The data might have chunks created in a specific order, like an English section followed by others. To capture this slice, we use the `sample` method.


In [5]:
# Option to display all columns
pd.set_option('display.max_columns', None)
survey_df.sample(5)

Timestamp   Age Primary streaming service  Hours per day  \
669  9/15/2022 15:30:41  20.0                   Spotify            1.0   
33   8/28/2022 10:59:53  17.0                   Spotify            4.0   
549   9/3/2022 16:00:31  18.0                   Spotify            2.0   
199  8/28/2022 20:57:57  28.0             YouTube Music            2.0   
264   8/29/2022 0:39:59  19.0                   Spotify            2.0   

    While working Instrumentalist Composer Fav genre Exploratory  \
669           Yes              No       No     K pop         Yes   
33             No              No       No      Rock         Yes   
549            No             Yes       No      Rock         Yes   
199           Yes              No       No      Rock         Yes   
264           Yes              No       No   Country         Yes   

    Foreign languages    BPM Frequency [Classical] Frequency [Country]  \
669               Yes  120.0             Sometimes               Never   
33                Yes  142.0                Rarely              Rarely   
549                No  115.0                 Never               Never   
199                No  172.0                 Never              Rarely   
264                No  125.0                Rarely     Very frequently   

    Frequency [EDM] Frequency [Folk] Frequency [Gospel] Frequency [Hip hop]  \
669          Rarely            Never          Sometimes               Never   
33           Rarely  Very frequently              Never           Sometimes   
549           Never            Never              Never              Rarely   
199          Rarely           Rarely              Never           Sometimes   
264          Rarely        Sometimes             Rarely               Never   

    Frequency [Jazz] Frequency [K pop] Frequency [Latin] Frequency [Lofi]  \
669           Rarely   Very frequently             Never        Sometimes   
33            Rarely            Rarely             Never           Rarely   
549            Never             Never             Never            Never   
199            Never            Rarely             Never        Sometimes   
264            Never             Never            Rarely           Rarely   

    Frequency [Metal]  Frequency [Pop]  Frequency [R&B] Frequency [Rap]  \
669             Never  Very frequently        Sometimes           Never   
33             Rarely  Very frequently           Rarely       Sometimes   
549         Sometimes        Sometimes           Rarely          Rarely   
199            Rarely        Sometimes  Very frequently          Rarely   
264         Sometimes           Rarely            Never           Never   

    Frequency [Rock] Frequency [Video game music]  Anxiety  Depression  \
669           Rarely                        Never      8.0         7.0   
33   Very frequently                       Rarely      5.0         6.0   
549  Very frequently                        Never      8.0         3.0   
199  Very frequently                       Rarely     10.0         8.0   
264        Sometimes                        Never      7.0         2.0   

     Insomnia  OCD Music effects    Permissions  
669       1.0  1.0     No effect  I understand.  
33        6.0  1.0       Improve  I understand.  
549       1.0  3.0       Improve  I understand.  
199       6.0  2.0       Improve  I understand.  
264       0.0  7.0     No effect  I understand.

In [9]:
survey_df['Frequency [Classical]'].unique()

array(['Rarely', 'Sometimes', 'Never', 'Very frequently'], dtype=object)

We can further characterize our datasets using `info`


In [6]:
survey_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 736 entries, 0 to 735
Data columns (total 33 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   Timestamp                     736 non-null    object 
 1   Age                           735 non-null    float64
 2   Primary streaming service     735 non-null    object 
 3   Hours per day                 736 non-null    float64
 4   While working                 733 non-null    object 
 5   Instrumentalist               732 non-null    object 
 6   Composer                      735 non-null    object 
 7   Fav genre                     736 non-null    object 
 8   Exploratory                   736 non-null    object 
 9   Foreign languages             732 non-null    object 
 10  BPM                           629 non-null    float64
 11  Frequency [Classical]         736 non-null    object 
 12  Frequency [Country]           736 non-null    object 
 13  Frequ

This method provides a concise summary of our table. Here's a breakdown of the information it offers:

The index comprises 736 entries, ranging from 0 to 735.

A brief description of each column is printed, including its non-null count and data type (dtype).

- **Non-null count** indicates the number of non-missing values in each column. In a table of this size, we typically wouldn't expect null values. However, understanding the presence of missing values is crucial for assessing the completeness of our data.
- **dtype** denotes the *type of data* present in each column. A comprehensive list of available data types is provided below. Ensuring accurate data types enables efficient operations and allows us to leverage pre-coded manipulations for data analysis.

| Pandas dtype    | What It Does                       | Common Operations                        | Advantages                                      | Disadvantages                                |
|-----------------|------------------------------------|------------------------------------------|-------------------------------------------------|----------------------------------------------|
| `object`        | Stores mixed types, often strings  | String operations, `.str` accessor       | Can hold mixed types                            | Inefficient, not suitable for numerical ops  |
| `int64`         | Integer numbers                    | Mathematical operations                  | Memory efficient, fast operations               | Cannot represent `NaN` in versions < 1.0     |
| `float64`       | Floating-point numbers             | Mathematical operations                  | Can represent decimals and `NaN`                | Less memory efficient than `int64`           |
| `bool`          | Boolean values (True/False)        | Logical operations                       | Very memory efficient                           | Limited to boolean logic                     |
| `datetime64[ns]`| Date and time values               | Date/time operations, `.dt` accessor     | Precise, timezone-aware operations             | More memory than simpler types               |
| `timedelta[ns]` | Differences between time points    | Time deltas, `.dt` accessor              | Precise calculations of time spans             | More complex to understand and manipulate    |
| `category`      | Finite list of text values         | Statistical ops on categorical data, `.cat` accessor | Memory efficient for small sets of unique values | Not suitable for large number of unique values |
| `complex`       | Complex numbers with real and imaginary parts | Mathematical operations on complex numbers | Can represent complex numbers                    | Rarely used, not supported by all libraries  |

As we can see in the previous table, our column `timestamp` has a dtype of object. This means that all its values are strings that contain timestamps. Storing data information this way isn't very efficient, as we can observe in the next cell.

In [7]:
# Example of how difficult it is to extract the month from a string column: 
survey_df['Timestamp'].str.split('/').str.get(0)

0       8
1       8
2       8
3       8
4       8
       ..
731    10
732    11
733    11
734    11
735    11
Name: Timestamp, Length: 736, dtype: object

This is not to mention that dates can sometimes be in various formats, such as `2023-09-21`, where the previous expression wouldn't capture all the years.

Now, we transform the column into a datetime type and retrieve the month in a simpler way. Although it seems straightforward in this example, the issue with differently formatted dates persists. It is almost always better to convert it to this format, as the time it takes for a column to be preprocessed is usually much smaller than the time this conversion will save us.

In [8]:
# Compared to using a datetime datatype
pd.to_datetime(survey_df['Timestamp']).dt.month

0       8
1       8
2       8
3       8
4       8
       ..
731    10
732    11
733    11
734    11
735    11
Name: Timestamp, Length: 736, dtype: int64

# Inplace and out-of-place operators.

Applying operators to isolated series or dataframes doesn't modify the object; instead, it makes a copy of the object and then applies operators to it. Some methods have an 'inplace' argument (a flag with specific values) that can update the data directly. Alternatively, you can reassign values to keep them updated.

In the previous cell, we changed the dtype of 'release_date' to datetime. However, when we check its dtype again, we find it's still an object.


In [9]:
# We can check the datatype of a column by just calling its values. 
# At the bottom we would find its dtype
survey_df['Timestamp']

0       8/27/2022 19:29:02
1       8/27/2022 19:57:31
2       8/27/2022 21:28:18
3       8/27/2022 21:40:40
4       8/27/2022 21:54:47
              ...         
731    10/30/2022 14:37:28
732     11/1/2022 22:26:42
733     11/3/2022 23:24:38
734     11/4/2022 17:31:47
735      11/9/2022 1:55:20
Name: Timestamp, Length: 736, dtype: object

In [10]:
# We can also find its dtype directly by accessing the dataframe's property
survey_df.dtypes


Timestamp                        object
Age                             float64
Primary streaming service        object
Hours per day                   float64
While working                    object
Instrumentalist                  object
Composer                         object
Fav genre                        object
Exploratory                      object
Foreign languages                object
BPM                             float64
Frequency [Classical]            object
Frequency [Country]              object
Frequency [EDM]                  object
Frequency [Folk]                 object
Frequency [Gospel]               object
Frequency [Hip hop]              object
Frequency [Jazz]                 object
Frequency [K pop]                object
Frequency [Latin]                object
Frequency [Lofi]                 object
Frequency [Metal]                object
Frequency [Pop]                  object
Frequency [R&B]                  object
Frequency [Rap]                  object


To make the change persistent, we need to assign it to the previous dataframe

In [11]:
survey_df['Timestamp']=pd.to_datetime(survey_df['Timestamp'])

survey_df.dtypes

Timestamp                       datetime64[ns]
Age                                    float64
Primary streaming service               object
Hours per day                          float64
While working                           object
Instrumentalist                         object
Composer                                object
Fav genre                               object
Exploratory                             object
Foreign languages                       object
BPM                                    float64
Frequency [Classical]                   object
Frequency [Country]                     object
Frequency [EDM]                         object
Frequency [Folk]                        object
Frequency [Gospel]                      object
Frequency [Hip hop]                     object
Frequency [Jazz]                        object
Frequency [K pop]                       object
Frequency [Latin]                       object
Frequency [Lofi]                        object
Frequency [Me

# The importance of an index:
- Allows for fast lookups
- Convenient data alignment
- Meaningful labeling

If there is a column that has unique values, it is convenient to assign it to an index so that each row is uniquely identified. We can see the unique values of each column using nunique.


In [12]:
len(survey_df)

736

In [13]:
survey_df.nunique().sort_values()

Permissions                       1
Foreign languages                 2
Composer                          2
Instrumentalist                   2
Exploratory                       2
While working                     2
Music effects                     3
Frequency [Rap]                   4
Frequency [Rock]                  4
Frequency [R&B]                   4
Frequency [Jazz]                  4
Frequency [Video game music]      4
Frequency [Pop]                   4
Frequency [Metal]                 4
Frequency [Lofi]                  4
Frequency [Latin]                 4
Frequency [K pop]                 4
Frequency [Hip hop]               4
Frequency [Folk]                  4
Frequency [EDM]                   4
Frequency [Country]               4
Frequency [Classical]             4
Frequency [Gospel]                4
Primary streaming service         6
Anxiety                          12
Depression                       12
Insomnia                         12
OCD                         

In [14]:
survey_df['Timestamp'].value_counts()

2022-08-28 16:15:08    2
2022-08-27 19:29:02    1
2022-09-01 21:07:33    1
2022-09-01 19:09:32    1
2022-09-01 19:36:54    1
                      ..
2022-08-28 23:34:19    1
2022-08-28 23:34:37    1
2022-08-28 23:40:54    1
2022-08-28 23:42:24    1
2022-11-09 01:55:20    1
Name: Timestamp, Length: 735, dtype: int64

Only `timestamp` is close to being an index, but it repeats, indicating simultaneous annotations. An index is crucial for faster and robust data analysis. Key characteristics of a good index:

- Uniqueness: Each record is uniquely identified, avoiding ambiguity.
- Immutability: Values in the index column remain constant, ensuring data integrity.
- Sequential order: A natural order improves query performance for retrievals and displays.
- Relevance: The index should be pertinent to frequent analyses, allowing quick access to crucial data.
- Minimal size: Compact indexes conserve memory and accelerate data operations.
- Performance: Optimize the index for common use cases, enhancing database or DataFrame performance.


Since we have numeric columns the method `describe` gives us a quick snapshot of their contents

In [18]:
survey_df.describe()

Age  Hours per day           BPM     Anxiety  Depression  \
count  735.000000     736.000000  6.290000e+02  736.000000  736.000000   
mean    25.206803       3.572758  1.589948e+06    5.837636    4.796196   
std     12.054970       3.028199  3.987261e+07    2.793054    3.028870   
min     10.000000       0.000000  0.000000e+00    0.000000    0.000000   
25%     18.000000       2.000000  1.000000e+02    4.000000    2.000000   
50%     21.000000       3.000000  1.200000e+02    6.000000    5.000000   
75%     28.000000       5.000000  1.440000e+02    8.000000    7.000000   
max     89.000000      24.000000  1.000000e+09   10.000000   10.000000   

         Insomnia         OCD  
count  736.000000  736.000000  
mean     3.738451    2.637228  
std      3.088689    2.842017  
min      0.000000    0.000000  
25%      1.000000    0.000000  
50%      3.000000    2.000000  
75%      6.000000    5.000000  
max     10.000000   10.000000

Here we can see that the mean age is 25, and its median is at 21, indicating a slightly skewed distribution. We also observe a higher mean score on anxiety, which may be an indication of where to look for interesting insights.

# Changing dtypes

In the Frequency column, all the values are of type 'object,' indicating they are strings. However, this data is clearly ordinal, with the value `Never` being less than the value `Sometimes`. We can replace these values to better reflect the type of data we have.


In [19]:
survey_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 736 entries, 0 to 735
Data columns (total 33 columns):
 #   Column                        Non-Null Count  Dtype         
---  ------                        --------------  -----         
 0   Timestamp                     736 non-null    datetime64[ns]
 1   Age                           735 non-null    float64       
 2   Primary streaming service     735 non-null    object        
 3   Hours per day                 736 non-null    float64       
 4   While working                 733 non-null    object        
 5   Instrumentalist               732 non-null    object        
 6   Composer                      735 non-null    object        
 7   Fav genre                     736 non-null    object        
 8   Exploratory                   736 non-null    object        
 9   Foreign languages             732 non-null    object        
 10  BPM                           629 non-null    float64       
 11  Frequency [Classical]         73

First, we select all the columns that start with the word `Frequency` by accessing the str property of the columns series. This also highlights the importance of having the correct dtype for each of the series.

In [20]:
survey_df.loc[:,survey_df.columns.str.startswith('Frequency')]

Frequency [Classical] Frequency [Country]  Frequency [EDM]  \
0                  Rarely               Never           Rarely   
1               Sometimes               Never            Never   
2                   Never               Never  Very frequently   
3               Sometimes               Never            Never   
4                   Never               Never           Rarely   
..                    ...                 ...              ...   
731       Very frequently              Rarely            Never   
732                Rarely              Rarely            Never   
733                Rarely           Sometimes        Sometimes   
734       Very frequently               Never            Never   
735             Sometimes              Rarely  Very frequently   

    Frequency [Folk] Frequency [Gospel] Frequency [Hip hop] Frequency [Jazz]  \
0              Never              Never           Sometimes            Never   
1             Rarely          Sometimes              Rarely  Very frequently   
2              Never              Never              Rarely           Rarely   
3             Rarely          Sometimes               Never  Very frequently   
4              Never             Rarely     Very frequently            Never   
..               ...                ...                 ...              ...   
731        Sometimes              Never           Sometimes           Rarely   
732            Never              Never               Never           Rarely   
733           Rarely             Rarely     Very frequently           Rarely   
734            Never              Never               Never           Rarely   
735        Sometimes             Rarely     Very frequently  Very frequently   

    Frequency [K pop] Frequency [Latin] Frequency [Lofi] Frequency [Metal]  \
0     Very frequently   Very frequently           Rarely             Never   
1              Rarely         Sometimes           Rarely             Never   
2     Very frequently             Never        Sometimes         Sometimes   
3           Sometimes   Very frequently        Sometimes             Never   
4     Very frequently         Sometimes        Sometimes             Never   
..                ...               ...              ...               ...   
731             Never         Sometimes           Rarely            Rarely   
732             Never             Never           Rarely             Never   
733            Rarely            Rarely        Sometimes            Rarely   
734             Never             Never            Never             Never   
735         Sometimes             Never           Rarely             Never   

     Frequency [Pop]  Frequency [R&B]  Frequency [Rap] Frequency [Rock]  \
0    Very frequently        Sometimes  Very frequently            Never   
1          Sometimes        Sometimes           Rarely  Very frequently   
2             Rarely            Never           Rarely           Rarely   
3          Sometimes        Sometimes            Never            Never   
4          Sometimes  Very frequently  Very frequently            Never   
..               ...              ...              ...              ...   
731  Very frequently            Never           Rarely  Very frequently   
732  Very frequently            Never            Never        Sometimes   
733        Sometimes        Sometimes        Sometimes           Rarely   
734            Never            Never            Never            Never   
735        Sometimes  Very frequently  Very frequently  Very frequently   

    Frequency [Video game music]  
0                      Sometimes  
1                         Rarely  
2                Very frequently  
3                          Never  
4                         Rarely  
..                           ...  
731                        Never  
732                    Sometimes  
733                       Rarely  
734                    Sometimes  
735                       R

First, we observe all the different values in the columns, and then we will replace these different values using a dictionary and the replace method.

In [21]:
freq_columns=survey_df.columns.str.startswith('Frequency')
freq_columns

array([False, False, False, False, False, False, False, False, False,
       False, False,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
       False, False, False, False, False, False])

All of them have the same values. Now, we replace the values using a dictionary. We are allowing a space of one between them, but the difference can be greater.

In [22]:
dict_replace={'Never':0,'Rarely':1,'Sometimes':2,'Very frequently':3}

survey_df.loc[:,freq_columns].replace(dict_replace)

Frequency [Classical]  Frequency [Country]  Frequency [EDM]  \
0                        1                    0                1   
1                        2                    0                0   
2                        0                    0                3   
3                        2                    0                0   
4                        0                    0                1   
..                     ...                  ...              ...   
731                      3                    1                0   
732                      1                    1                0   
733                      1                    2                2   
734                      3                    0                0   
735                      2                    1                3   

     Frequency [Folk]  Frequency [Gospel]  Frequency [Hip hop]  \
0                   0                   0                    2   
1                   1                   2                    1   
2                   0                   0                    1   
3                   1                   2                    0   
4                   0                   1                    3   
..                ...                 ...                  ...   
731                 2                   0                    2   
732                 0                   0                    0   
733                 1                   1                    3   
734                 0                   0                    0   
735                 2                   1                    3   

     Frequency [Jazz]  Frequency [K pop]  Frequency [Latin]  Frequency [Lofi]  \
0                   0                  3                  3                 1   
1                   3                  1                  2                 1   
2                   1                  3                  0                 2   
3                   3                  2                  3                 2   
4                   0                  3                  2                 2   
..                ...                ...                ...               ...   
731                 1                  0                  2                 1   
732                 1                  0                  0                 1   
733                 1                  1                  1                 2   
734                 1                  0                  0                 0   
735                 3                  2                  0                 1   

     Frequency [Metal]  Frequency [Pop]  Frequency [R&B]  Frequency [Rap]  \
0                    0                3                2                3   
1                    0                2                2                1   
2                    2                1                0                1   
3                    0                2                2                0   
4                    0                2                3                3   
..                 ...              ...              ...              ...   
731                  1                3                0                1   
732                  0                3                0                0   
733                  1                2                2                2   
734                  0                0                0                0   
735                  0                2                3                3   

     Frequency [Rock]  Frequency [Video game music]  
0                   0                             2  
1                   3                             1  
2                   1                             3  
3                   0                             0  
4                   0                             1  
..                ...                           ...  
731                 3                             0  
732                 2                             2  
733        

# Traversing the DataFrame
In the vast sea of data within a dataframe, finding the information you need can be like searching for a needle in a haystack, especially with lengthy tables. Fortunately, pandas provides an array of methods to streamline this process. Let's explore some of the most prevalent and valuable techniques for extracting the desired information.

## `loc`

- Purpose: Accesses a group of rows and columns by labels or boolean array.
- Usage: `df.loc[row_label, column_label]`
- Useful When: You know the label of the index or columns.

In [30]:
# Access the data for the third respondent
third_respondent = survey_df.loc[2]
third_respondent

Timestamp                       2022-08-27 21:28:18
Age                                            18.0
Primary streaming service                   Spotify
Hours per day                                   4.0
While working                                    No
Instrumentalist                                  No
Composer                                         No
Fav genre                          Video game music
Exploratory                                      No
Foreign languages                               Yes
BPM                                           132.0
Frequency [Classical]                           0.0
Frequency [Country]                             0.0
Frequency [EDM]                                 3.0
Frequency [Folk]                                0.0
Frequency [Gospel]                              0.0
Frequency [Hip hop]                             1.0
Frequency [Jazz]                                1.0
Frequency [K pop]                               3.0
Frequency [L

The output is the third register in the df (remember in Python we start at zero). Since the index was just a number we could access it this way, but in other dfs we will have to put the index value directly, like we do in the following cell with the column name `Fav genre`

In [31]:
third_respondent_fav_genre = survey_df.loc[2,'Fav genre']
third_respondent_fav_genre

'Video game music'

## `iloc`
- Purpose: Accesses a group of rows and columns by integer position(s).
- Usage: `df.iloc[row_index, column_index]`
- Useful When: You want to access elements based on their position (integer index).

In [32]:
# Access the data for the a slice of the second to the fifth register and second to sixth coulmn
second_to_fifth = survey_df.iloc[2:5,2:6]
second_to_fifth

Primary streaming service  Hours per day While working Instrumentalist
2                   Spotify            4.0            No              No
3             YouTube Music            2.5           Yes              No
4                   Spotify            4.0           Yes              No

The operation we performed in the previous cell is called a **slicing**. We specify ranges of rows and columns we want to have. This is useful when applying operations only to those specific rows and columns

## `isin`
- Purpose: Filter data based on a list of values.
- Usage: `df[column].isin(values)`
- Useful When: Filtering data by a list of categorical values.

In [34]:
# Select respondents who use specific streaming services
services = ['Spotify', 'Apple Music']
filtered_df = survey_df[survey_df['Primary streaming service'].
                        isin(services)]
filtered_df

Timestamp   Age Primary streaming service  Hours per day  \
0   2022-08-27 19:29:02  18.0                   Spotify            3.0   
2   2022-08-27 21:28:18  18.0                   Spotify            4.0   
4   2022-08-27 21:54:47  18.0                   Spotify            4.0   
5   2022-08-27 21:56:50  18.0                   Spotify            5.0   
7   2022-08-27 22:18:59  21.0                   Spotify            1.0   
..                  ...   ...                       ...            ...   
726 2022-10-23 20:50:27  18.0               Apple Music           18.0   
730 2022-10-30 13:15:26  21.0                   Spotify            2.0   
731 2022-10-30 14:37:28  17.0                   Spotify            2.0   
732 2022-11-01 22:26:42  18.0                   Spotify            1.0   
734 2022-11-04 17:31:47  19.0                   Spotify            5.0   

    While working Instrumentalist Composer         Fav genre Exploratory  \
0             Yes             Yes      Yes             Latin         Yes   
2              No              No       No  Video game music          No   
4             Yes              No       No               R&B         Yes   
5             Yes             Yes      Yes              Jazz         Yes   
7             Yes              No       No             K pop         Yes   
..            ...             ...      ...               ...         ...   
726           Yes              No       No               EDM         Yes   
730           Yes              No       No               R&B         Yes   
731           Yes             Yes       No              Rock         Yes   
732           Yes             Yes       No               Pop         Yes   
734           Yes             Yes       No         Classical          No   

    Foreign languages    BPM  Frequency [Classical]  Frequency [Country]  \
0                 Yes  156.0                    1.0                  0.0   
2                 Yes  132.0                    0.0                  0.0   
4                  No  107.0                    0.0                  0.0   
5                 Yes   86.0                    1.0                  2.0   
7                 Yes   95.0                    0.0                  0.0   
..                ...    ...                    ...                  ...   
726                No   90.0                    2.0                  1.0   
730               Yes  147.0                    2.0                  0.0   
731               Yes  120.0                    3.0                  1.0   
732               Yes  160.0                    1.0                  1.0   
734                No  170.0                    3.0                  0.0   

     Frequency [EDM]  Frequency [Folk]  Frequency [Gospel]  \
0                1.0               0.0                 0.0   
2                3.0               0.0                 0.0   
4                1.0               0.0                 1.0   
5                0.0               0.0                 0.0   
7                1.0               0.0                 0.0   
..               ...               ...                 ...   
726              3.0               0.0                 1.0   
730              2.0               1.0                 0.0   
731              0.0               2.0                 0.0   
732              0.0               0.0                 0.0   
734              0.0               0.0                 0.0   

     Frequency [Hip hop]  Frequency [Jazz]  Frequency [K pop]  \
0                    2.0               0.0                3.0   
2                    1.0               1.0                3.0   
4                    3.0               0.0                3.0   
5                    2.0               3.0                3.0   
7                    3.0               1.0                3.0   
..                   ...               ...                ...   
726                  2.0               2.0                0.0   
730                  0.0          

## Masking
- Purpose: Selecting data based on a condition.
- Usage: `df[mask]` where mask is a boolean array.
- Useful When: You need to filter data based on a specific condition.

In [37]:
# Get a boolean pd.Series respondents aged over 30
over_30 = survey_df['Age'] > 30
over_30

0      False
1       True
2      False
3       True
4      False
       ...  
731    False
732    False
733    False
734    False
735    False
Name: Age, Length: 736, dtype: bool

Now we can select only the values where the boolean series is True by using the following notation

`df_masked=df[mask]`

note that both the mask and the df will have to have the same indexes.

In [38]:
# Select respondents aged over 30
df_over_30 = survey_df[survey_df['Age'] > 30]
df_over_30

Timestamp   Age          Primary streaming service  \
1   2022-08-27 19:57:31  63.0                            Pandora   
3   2022-08-27 21:40:40  61.0                      YouTube Music   
42  2022-08-28 11:55:54  37.0                      YouTube Music   
46  2022-08-28 12:15:02  36.0                            Spotify   
57  2022-08-28 13:08:24  42.0  I do not use a streaming service.   
..                  ...   ...                                ...   
679 2022-09-19 23:53:37  36.0                      YouTube Music   
695 2022-09-28 17:25:48  89.0                            Spotify   
702 2022-10-03 10:53:28  37.0                            Spotify   
703 2022-10-03 11:30:10  44.0                      YouTube Music   
709 2022-10-04 00:48:34  35.0                            Spotify   

     Hours per day While working Instrumentalist Composer         Fav genre  \
1             1.50           Yes              No       No              Rock   
3             2.50           Yes              No      Yes              Jazz   
42            0.25            No              No       No  Video game music   
46            1.00            No             Yes      Yes         Classical   
57            0.00           NaN              No       No              Rock   
..             ...           ...             ...      ...               ...   
679           3.00           Yes              No       No           Country   
695          24.00           Yes             Yes      Yes               Rap   
702           2.00           Yes              No       No              Folk   
703           1.50            No              No       No              Folk   
709           4.00           Yes             Yes      Yes           Hip hop   

    Exploratory Foreign languages    BPM  Frequency [Classical]  \
1           Yes                No  119.0                    2.0   
3           Yes               Yes   84.0                    2.0   
42          Yes                No  147.0                    3.0   
46           No                No  132.0                    3.0   
57           No               Yes    NaN                    1.0   
..          ...               ...    ...                    ...   
679          No               Yes  105.0                    3.0   
695          No                No  143.0                    0.0   
702         Yes                No  186.0                    2.0   
703          No                No   99.0                    1.0   
709         Yes                No   90.0                    2.0   

     Frequency [Country]  Frequency [EDM]  Frequency [Folk]  \
1                    0.0              0.0               1.0   
3                    0.0              0.0               1.0   
42                   1.0              0.0               1.0   
46                   2.0              0.0               1.0   
57                   1.0              0.0               0.0   
..                   ...              ...               ...   
679                  3.0              2.0               0.0   
695                  0.0              1.0               1.0   
702                  2.0              0.0               2.0   
703                  3.0              0.0               3.0   
709                  1.0              1.0               2.0   

     Frequency [Gospel]  Frequency [Hip hop]  Frequency [Jazz]  \
1                   2.0                  1.0               3.0   
3                   2.0                  0.0               3.0   
42                  0.0                  1.0               2.0   
46                  0.0                  0.0               0.0   
57                  0.0                  0.0               1.0   
..                  ...                  ...               ...   
679                 0.0                  1.0               0.0   
695                 0.0                  3.0               2.0   
702                 0.0                  1.0               2.0   
703                 1.0   

When masking multiple conditions you need to wrap each of them in a parenthesis and the logic operators change form `and` to `&` and from `or` to `|`. As we can see in the following example

In [42]:
streaming_services=['Pandora', 'YouTube Music']
df_over_30_pandora_yt = survey_df[(survey_df['Age'] > 30) & 
    (survey_df['Primary streaming service'].isin(streaming_services))]
df_over_30_pandora_yt.head()

Timestamp   Age Primary streaming service  Hours per day  \
1   2022-08-27 19:57:31  63.0                   Pandora           1.50   
3   2022-08-27 21:40:40  61.0             YouTube Music           2.50   
42  2022-08-28 11:55:54  37.0             YouTube Music           0.25   
93  2022-08-28 14:51:02  36.0             YouTube Music           1.00   
158 2022-08-28 18:57:39  41.0             YouTube Music           4.00   

    While working Instrumentalist Composer         Fav genre Exploratory  \
1             Yes              No       No              Rock         Yes   
3             Yes              No      Yes              Jazz         Yes   
42             No              No       No  Video game music         Yes   
93            Yes              No       No              Rock         Yes   
158           Yes              No       No               Pop         Yes   

    Foreign languages    BPM  Frequency [Classical]  Frequency [Country]  \
1                  No  119.0                    2.0                  0.0   
3                 Yes   84.0                    2.0                  0.0   
42                 No  147.0                    3.0                  1.0   
93                 No  110.0                    2.0                  1.0   
158                No   60.0                    1.0                  0.0   

     Frequency [EDM]  Frequency [Folk]  Frequency [Gospel]  \
1                0.0               1.0                 2.0   
3                0.0               1.0                 2.0   
42               0.0               1.0                 0.0   
93               0.0               0.0                 0.0   
158              2.0               3.0                 0.0   

     Frequency [Hip hop]  Frequency [Jazz]  Frequency [K pop]  \
1                    1.0               3.0                1.0   
3                    0.0               3.0                2.0   
42                   1.0               2.0                0.0   
93                   0.0               0.0                0.0   
158                  3.0               2.0                1.0   

     Frequency [Latin]  Frequency [Lofi]  Frequency [Metal]  Frequency [Pop]  \
1                  2.0               1.0                0.0              2.0   
3                  3.0               2.0                0.0              2.0   
42                 1.0               0.0                0.0              3.0   
93                 0.0               0.0                1.0              2.0   
158                1.0               2.0                1.0              3.0   

     Frequency [R&B]  Frequency [Rap]  Frequency [Rock]  \
1                2.0              1.0               3.0   
3                2.0              0.0               0.0   
42               2.0              0.0               2.0   
93               1.0              1.0               3.0   
158              3.0              2.0               3.0   

     Frequency [Video game music]  Anxiety  Depression  Insomnia  OCD  \
1                             1.0      7.0         2.0       2.0  1.0   
3                             0.0      9.0         7.0       3.0  3.0   
42                            2.0      3.0         2.0      10.0  0.0   
93                            0.0      6.0         2.0       3.0  1.0   
158                           1.0      7.0         2.0       0.0  7.0   

    Music effects    Permissions  
1             NaN  I understand.  
3         Improve  I understand.  
42      No effect  I understand.  
93        Improve  I understand.  
158       Improve  I understand.

## Querying:
- Purpose: Query the columns of a DataFrame with a boolean expression.
- Usage: `df.query('expression')`
- Useful When: You want to filter data using a concise, readable syntax.

In [48]:
# Select respondents who spend more than 3 hours per day and are over 25 years old
specific_group = survey_df.query('`Hours per day` > 3 & Age > 25')
specific_group

Timestamp   Age          Primary streaming service  \
64  2022-08-28 13:42:20  32.0                            Spotify   
65  2022-08-28 13:43:06  36.0                            Spotify   
67  2022-08-28 13:48:11  29.0                            Spotify   
90  2022-08-28 14:45:44  30.0                            Spotify   
91  2022-08-28 14:47:08  43.0  I do not use a streaming service.   
..                  ...   ...                                ...   
635 2022-09-12 19:13:06  49.0  I do not use a streaming service.   
651 2022-09-13 08:05:59  26.0                            Spotify   
673 2022-09-18 16:23:29  27.0                            Spotify   
695 2022-09-28 17:25:48  89.0                            Spotify   
709 2022-10-04 00:48:34  35.0                            Spotify   

     Hours per day While working Instrumentalist Composer         Fav genre  \
64             5.0           Yes              No       No              Rock   
65             6.0           Yes              No       No             Metal   
67             4.0            No              No       No           Hip hop   
90             5.0           Yes             Yes       No             Metal   
91             6.0           Yes             Yes      Yes  Video game music   
..             ...           ...             ...      ...               ...   
635            8.0           Yes             Yes      Yes         Classical   
651            5.0           Yes             Yes       No              Folk   
673           10.0           Yes             Yes      Yes              Jazz   
695           24.0           Yes             Yes      Yes               Rap   
709            4.0           Yes             Yes      Yes           Hip hop   

    Exploratory Foreign languages    BPM  Frequency [Classical]  \
64           No               Yes   91.0                    2.0   
65          Yes               Yes  174.0                    3.0   
67          Yes                No  140.0                    1.0   
90          Yes                No  160.0                    2.0   
91          Yes               Yes  109.0                    2.0   
..          ...               ...    ...                    ...   
635          No                No   68.0                    3.0   
651         Yes               Yes   92.0                    2.0   
673         Yes               Yes  132.0                    3.0   
695          No                No  143.0                    0.0   
709         Yes                No   90.0                    2.0   

     Frequency [Country]  Frequency [EDM]  Frequency [Folk]  \
64                   0.0              1.0               2.0   
65                   0.0              1.0               3.0   
67                   2.0              1.0               0.0   
90                   1.0              0.0               2.0   
91                   1.0              3.0               1.0   
..                   ...              ...               ...   
635                  0.0              0.0               0.0   
651                  1.0              1.0               3.0   
673                  0.0              0.0               3.0   
695                  0.0              1.0               1.0   
709                  1.0              1.0               2.0   

     Frequency [Gospel]  Frequency [Hip hop]  Frequency [Jazz]  \
64                  0.0                  2.0               1.0   
65                  0.0                  0.0               0.0   
67                  1.0                  3.0               1.0   
90                  0.0                  1.0               0.0   
91                  1.0                  1.0               1.0   
..                  ...                  ...               ...   
635                 0.0                  0.0               1.0   
651                 0.0                  3.0               3.0   
673                 1.0                  1.0               3.0   
695                 0.0   

Note that we could have achieved the same thing using masking. The differences between these two methods are shown in the table below

| Aspect                      | Querying                            | Masking                             |
|-----------------------------|-------------------------------------|-------------------------------------|
| **Usage Scenario**          | When you need to filter data using a concise, readable string expression. Ideal for complex conditions. | When you want to filter data based on conditions directly written with DataFrame's columns. More Pythonic for simple conditions. |
| **Syntax Example**          | `df.query('Age > 30 & "Hours per day" > 3')` | `df[(df.Age > 30) & (df['Hours per day'] > 3)]` |
| **Computational Efficiency**| Generally faster for complex queries due to optimized internal parsing and execution. | Slightly slower for complex conditions due to direct evaluation in Python's interpreter. However, differences are often minimal for small to moderately sized DataFrames. |
| **In-Place / Out-of-Place** | Out-of-place. Returns a new DataFrame. | Out-of-place. Returns a new DataFrame. |
| **Readability**             | Better for complex conditions. Compact and closer to SQL-like syntax. | Better for simple conditions. Follows standard Python logical syntax. |
| **Flexibility**             | Limited to string expressions. Cannot use external variables directly without formatting. | More flexible. Allows the use of external Python variables and functions directly. |
| **Additional Notes**        | Requires string parsing. Might throw an error for invalid syntax. | Utilizes standard Python boolean indexing. Familiar to Python users. |


